### DISCUSSION TITLE
Fate of bokeh/bokeh docker images

### DISCUSSION THREAD

**Discussion by bryevdv** [bokeh/bokeh#12599] [Build / CI]
cc @ianthomas23 @Karel-van-de-Plassche 

Can https://hub.docker.com/r/bokeh/bokeh be deleted? The tidy-er in me would really like to. I'd really like the only images on dockerhub to be things that we are actively maintaining. it looks like there have been ~204 pulls in four years?

**Comment by Karel-van-de-Plassche** [comment]
If it's just me using it, for sure. I can move to the more actively maintained images if need be.

**Reply by ianthomas23** [reply]
It would be great if we could all use a single set of images as then we will have multiple eyes on them. If there are changes you need in the current approach, which is still experimental and subject to change anyway, then let us know.

**Reply by bryevdv** [reply]
@Karel-van-de-Plassche can you check out the `bokeh-dev` images @ianthomas23 has started pushing manually so far when you have a chance? These will have an automated job to update and be used in CI testing. If they seem workable for you then I will delete the older ones.

**Reply by Karel-van-de-Plassche** [reply]
Just to be sure, you are referring to the images pushed here: https://hub.docker.com/r/bokeh/bokeh-dev/tags. Meaning; I do not need to test docker-in-docker and follow the procedure outlined in https://github.com/bokeh/bokeh/wiki/Docker-dev-container#workflow. At least as first step?

**Reply by ianthomas23** [reply]
Yes to the images, no to the procedure. The procedure is a bit out of date, I was going to rewrite it when I merge the current PR which I haven't quite finished yet.

**Reply by Karel-van-de-Plassche** [reply]
Initial testing shows that the current `ENTRYPOINT` and `CMD` might need an update to use it in other CI; The current ENTRYPOINT assumes `/bokeh` is locally available, which in most CI, it will not be, as `git clone some_bokeh_repo.git` will be done _after_ the bokeh Docker image is loaded (probably).

Starting with Docker in Docker (https://docs.gitlab.com/ee/ci/docker/using_docker_build.html) and/or pre-test jobscripts.

**Reply by ianthomas23** [reply]
Note that the intention of this docker container is for developers, hence the `bokeh-dev` name. The idea is to give developers a way to build and test branches in a controlled way without having to worry about installing all the dependencies which can be problematic on different OSes/hardware. So the workflow is to start with the git repo on the host machine as any developer normally would, checkout the branch of interest then run the docker container using a script contained within the branch (hence the repo has to be present). The first time it is run it will install an appropriate conda environment and drop back to a bash shell for the developer to work as normal. This should all be achieved as easily as possible for the developer (i.e. running a single command) without them having to know anything much about docker. The developer can leave and restart the container without having to reinstall the conda environment and hence continue working very quickly.

The current PR also uses the docker container in a couple of CI runs, because, well, why not. This will probably mostly be used to check that the container can still run the tests, i.e. it is primarily validation of the container not the tests. Personally I am not particularly interested in using it in CI, so if this becomes problematic I will be very happy to drop it and just use it for development as originally planned.

If you'd like to use it for other CI then feel free to make whatever modifications you need, provided it doesn't interfere with the primary use case.

**Reply by Karel-van-de-Plassche** [reply]
Thanks for the explanation, very clear! Using this information and some Docker magic, I've set up a relatively simple `.gitlab-ci.yml` to run my own CI on gitlab.com, see https://gitlab.com/qualikiz-group/QuaLiKiz-dataslicer/-/merge_requests/1 and https://gitlab.com/qualikiz-group/QuaLiKiz-dataslicer/-/pipelines/695248066. This allows people with one foot in dev and one foot in user land to re-use the bokeh-dev image for their own development. Possible improvements are noted in the MR:

- Use Docker-in-Docker to build images _based on_ the `bokeh-dev` images on curated tags

This might save you 3-5 minutes of set up time every time the CI is ran.

**Reply by Karel-van-de-Plassche** [reply]
Extra: I did extend the docker-in-docker strategy here: https://gitlab.com/qualikiz-group/bokeh-ion-rangeslider/container_registry

**Reply by bryevdv** [reply]
@Karel-van-de-Plassche just for my own understanding, it sounds like the new container *will* work for you? If so please just give me a final 👍 when I can clean up docker hub

**Reply by Karel-van-de-Plassche** [reply]
Indeed it will! I'll be working on making a "bokeh plugin" (https://gitlab.com/qualikiz-group/bokeh-ion-rangeslider/container_registry) as example as well!
